### PR TITLE
[ML] Detect ML nodes by AZ from consistent state during rebalance

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
This commit ensures the ML nodes by availability zone considered during rebalancing trained model deployments is consistent and based off a specific cluster state.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NodeAvailabilityZoneMapper.java [hunk 1]
```diff
@@ -107,30 +108,33 @@ public OptionalInt getNumMlAvailabilityZones() {
...
-    private synchronized void buildNodesByAvailabilityZone() {
+    private synchronized void updateNodesByAvailabilityZone() {
         if (lastDiscoveryNodes == null) {
             allNodesByAvailabilityZone = Map.of();
             mlNodesByAvailabilityZone = allNodesByAvailabilityZone;
             return;
         }
+        NodesByAvailabilityZone nodesByAvailabilityZone = buildNodesByAvailabilityZone(lastDiscoveryNodes, awarenessAttributes);
+        this.allNodesByAvailabilityZone = nodesByAvailabilityZone.allNodes;
+        this.mlNodesByAvailabilityZone = nodesByAvailabilityZone.mlNodes;
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NodeAvailabilityZoneMapper.java:123]: Are the member variables still immutable?  Before the `Map.copyOf` enforced immutability.  But now are these just hash maps that could be modified?

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NodeAvailabilityZoneMapper.java:123]: I'm now using `Map.copyOf` when I create a new `NodesByAvailabilityZone` object.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NodeAvailabilityZoneMapper.java [hunk 2]
```diff
@@ -156,7 +160,27 @@ private synchronized void buildNodesByAvailabilityZone() {
...
+     *
+     * @param clusterState The cluster state whose nodes will be used to detect ML nodes by availability zone.
+     * @return A map whose keys are lists of awareness attribute values in the same order as the configured awareness attribute
+     *         names, and whose values are collections of nodes that have that combination of attributes. If availability zones
+     *         are not configured then the map will contain one entry mapping an empty list to a collection of all nodes. If
+     *         it is too early in the lifecycle of the node to know the answer then an empty map will be returned. An empty
+     *         map will also be returned if there are no ML nodes in the cluster. (These two empty map return scenarios can be
+     *         distinguished by calling one of the other methods.)
+     */
+    public synchronized Map<List<String>, Collection<DiscoveryNode>> buildMlNodesByAvailabilityZone(ClusterState clusterState) {
```
[the remark this answers is not in the document]
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NodeAvailabilityZoneMapper.java]: Good point. I've removed `synchronized` from that method.

**Review by droberts195 — APPROVED** [review]
LGTM

**Pull request merged**